### PR TITLE
Ref assemblies: Address more feedback on #17558

### PR DESCRIPTION
**Pull request by jcouv**
@tmat for review.

You had some additional feedback on handling nested types in `ConstantSignatureVisualizer` in `MetadataReaderUtils.cs`. I'll add that when I refine what goes into ref assemblies (at the moment, all types are included, regardless of accessibility).

Follows up on #17558

### CODE REVIEW

**Comment by jcouv** [comment]
@tmat Take a look when you get the chance, then I'll send more broadly for review.

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs [hunk 1]
```diff
@@ -393,19 +367,38 @@ internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, bool expec
...
-                        }
-                        else
-                        {
-                            AssertEx.NotEqual(throwNull, il);
-                        }
+                        ilValidator(il);
+                    }
+                    else
+                    {
+                        ilValidator(null);
```
Comment (tmat) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: > ilValidator(null); [](start = 23, length = 19)

Is it useful to call validator on non-existing bodies?

**Comment by tmat** [comment]
:shipit:

File: src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs (hunk 1 above)
Comment (jcouv) [inline src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs]: Generally, yes, I think so.
The purpose of the refactoring is to make this helper method more generic and re-usable. If I want to validate method bodies, I would want to see empty method bodies too. 
In my specific scenario, I don't actually care though.

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 2]
```diff
@@ -397,6 +397,34 @@ public void EmitMetadataOnly_DisallowOutputtingNetModule()
         }
 
         [Fact]
+        public void EmitMetadata_DisallowEmbedding()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: @tmat I'm having second thoughts about this specific case, which emits both a regular assembly and a ref assembly. It seems that we should allow embedding PDB information into the regular assembly in this case. What do you think?

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: Ah, yes. If we are emitting both then embedding should be allowed.

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: (only if we are emitting metadata only, without the full assembly being produced we should report error).

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by cston** [comment]
LGTM

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs [hunk 3]
```diff
@@ -2808,6 +2815,35 @@ private void PopulateStandaloneSignatures()
             }
         }
 
+        private int[] SerializeThrowNullMethodBodies(BlobBuilder ilBuilder, out Blob mvidStringFixup)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: >, out Blob mvidStringFixup [](start = 74, length = 26)

It doesn't look like this parameter is used in any meaningful way.

Comment (jcouv) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: It is only used to ensure that the caller's `mvidStringFixup` is initialized on this code path.

File: src/Compilers/VisualBasic/Portable/VBResources.resx [hunk 4]
```diff
@@ -5061,6 +5061,7 @@
 /target:winmdobj                  Create a Windows Metadata intermediate file
 /doc[+|-]                         Generates XML documentation file.
 /doc:&lt;file&gt;                       Generates XML documentation file to &lt;file&gt;.
+/refout:&lt;file&gt;                    Reference assembly output to generate
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/VBResources.resx]: Is this line going to align properly in the output?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -4402,6 +4402,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
  /target:winmdobj              Build a Windows Runtime intermediate file that 
                                is consumed by WinMDExp (Short form: /t:winmdobj)
  /doc:&lt;file&gt;                   XML Documentation file to generate
+ /refout:&lt;file&gt;                Reference assembly output to generate
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >/refout:&lt;file&gt;                Reference assembly output to generate [](start = 1, length = 73)

Is this line going to align properly in the output?

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: Could be done with a regular ```if``` statement though. i don't remember seen any other place in the code base where we add a parameter just for this purpose.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 5 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Why don't we use CDATA to avoid escaping XML entities?

**Comment by jcouv** [comment]
@AlekseyTs Yes, looks ok.

C#:
image
image

VB:
image
image

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 5 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: That would make sense, but that'll be for next time.

**Comment by AlekseyTs** [comment]
Alignment looks different in the two VB pictures.

**Comment by jcouv** [comment]
That's just an artifact of scaling of pictures on github. Notice that the new entries line up with their immediate and existing neighbors.

**Pull request merged**